src/ring_theory/unique_factorization_domain.lean

```diff
@@ -243,6 +249,16 @@ theorem prod_mono : ∀{a b : factor_set α}, a ≤ b → a.prod ≤ b.prod
 | a none h := show a.prod ≤ (⊤ : factor_set α).prod, by simp; exact le_top
 | (some a) (some b) h := prod_le_prod $ multiset.map_le_map $ with_top.coe_le_coe.1 $ h
 
+/-- `count p s` is the multiplicity of `p` in the factor_set `s`. -/
+def count [decidable_eq (associates α)] (p : { a : associates α // irreducible a }) :
```

Comment:
But don't we then have to write `count hp s`? Doing: def count [decidable_eq (associates α)] (p : associates α) {hp : irreducible p} doesn't seem to work...

Reply:
Indeed, the disadvantage is that it looks less like the mathematical notation. I'm less concerned about using standard notation than I am about keeping the proof obligations easier to manage. But then again, I am officially working in a computer science department. @jcommelin, what do you think?

If we prefer to have it look like `count p`, we could also go for something like:

```lean
def count_irr [decidable_eq (associates α)] (p : { a : associates α // irreducible a }) :=
/- the current definition -/

def count [decidable_eq (associates α)] (p : associates α) :=
if hp : irreducible p then count_irr (p, hp) else 0
```
(Not so sure about the names though, since both definitions only really work for irreducible elements.)

Reply:
I like this!
One problem I ran into though. In that second definition I'm getting a failed to synthesize type class instance for decidable (irreducible p). This is at the 'if'. Is there a decidable version of 'if'? Or is there some other way I can tell it to be classical? Adding noncomputable theory etc. at the top of the file fixes it, but that doesn't seem to be in the spirit of the file...

Reply:
I guess this works:

def bcount [decidable_eq (associates α)] (p : {a : associates α // irreducible a}) :
  factor_set α → ℕ
| none := 0
| (some s) := s.count p

section
open classical
local attribute [instance] prop_decidable

noncomputable def count' [decidable_eq (associates α)] (p : associates α) :
  factor_set α → ℕ :=
if hp : irreducible p then bcount ⟨p, hp⟩  else 0

end

Is there a better way? (bcount for "bundled count")

Reply:
Alternatively, you could add a `decidable (irreducible p)` variable:

```lean
variables [dec_eq : decidable_eq (associates α)]
include dec_eq

def bcount (p : {a : associates α // irreducible a}) :
factor_set α → ℕ
| none := 0
| (some s) := s.count p

variables [dec_irr : Π (p : associates α), decidable (irreducible p)]
include dec_irr

def count' (p : associates α) : factor_set α → ℕ :=
if hp : irreducible p then bcount ⟨p, hp⟩ else 0
variable [unique_factorization_domain α]
```

Do we want `p` to be an element of `associates α`, or just an `α`? The second case would look something like:

```lean
variables [dec_eq : decidable_eq (associates α)]
include dec_eq

def bcount (p : {a : associates α // irreducible a}) :
factor_set α → ℕ
| none := 0
| (some s) := s.count p

variables [dec_irr : Π (p : associates α), decidable (irreducible p)]
include dec_irr

def count' (p : α) : factor_set α → ℕ :=
if hp : irreducible (associates.mk p) then bcount ⟨associates.mk p, hp⟩ else 0
```

I don't have any good ideas which will turn out nicer, so my advice would be to pick one that looks nicest to you and try that out.

Reply:
I don't think we want to have `p` to be just an α, because we want to count `p` in the factor_set produced by `factor a` with `a :  associates α`, so I think it is more natural to have `p` be `associates α`.
I think I have something I'm happy with now. Thanks for the help!